### ISSUE TITLE
Migrate to new Architecture

### ISSUE THREAD

**Issue by birkskyum** [maplibre/maplibre-react-native#480]
## Motivation

We're currently at "react-native": "^0.74.6"

The New Architecture is default from 0.76

We can use the new arch though a compat layer #483, which can be removed by migrating the repo to 0.76.

**Intro to New Architecture**

In this new version async bridge is removed entirely.

This makes for better performance (lower overhead c++ <-> js), and allow sync+async, where prior is async only, which allow for some flattening of callbacks to sync code etc.

In addition the debugging experience is built around Chrome DevTools.

Read more here: https://reactnative.dev/blog/2024/10/23/the-new-architecture-is-here

**Comment by flohhhh** [comment]
Looking forward to this update :)

**Comment by KiwiKilian** [comment]
@birkskyum is this ticket to track progress on general support for the new architecture or migrating this project completely to the new architecture?

**Comment by birkskyum** [comment]
Tracking ticket for support of new architecture. I think it would be great to get the entire repo onto 0.76, but it's maybe possible to do in steps.

**Comment by birkskyum** [comment]
@KiwiKilian , I didn't know that there was a way to do a compat layer. I think we can close this with #483

**Comment by KiwiKilian** [comment]
Then you can also drop the 0.76 version reference – new arch is supported already quite a bit earlier. Also upgrading to this version is not exactly updating to the new arch.

#483 fixes issues, which prevented to make this library work in apps using the new architecture through the Interop Layer. This is fixed by adhering to [these guidelines](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/enable-libraries.md). 

The next step could be migrating the whole library to the new architecture. How to do it is described [here](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/enable-libraries-prerequisites.md). This also states:

> Before proceeding to convert your library to natively use TurboModules/Fabric, you should [ensure that it is compatible with the Interop Layer](https://github.com/reactwg/react-native-new-architecture/blob/main/docs/enable-libraries.md). There is no known performance cost imposed by the Interop Layer, but it will be removed at some point in the future. Today, it's safe and recommended to use the Interop Layer as a first solution to make your library compatible with the New Architecture quickly, and then you can consider migrating to the native APIs in the near future.

So there is no date yet available, when Interop Layer will be removed.

@birkskyum do you think switching to the new arch could be supported by a bounty? It looks like quite a bit of [work](https://github.com/rnmapbox/maps/pull/3048). Not sure if Software Mansion (big consultant/player in the React Native ecosystem) was contracted to do it, looks like they did most of the work.

**Comment by tyrauber** [comment]
We have a released 10.0.0.alpha.24 which includes an interop layer for the new react-native architecture, thanks to @KiwiKilian,  @rxnvee and @thibaultcapelli.  alpha.24 should be backwards compatible for those that want or need to use the old architecture, for both iOS and Android.  If you run into issues, you can lock at alpha.23, comment or open an ticket.

We were noticing some weird intermittent build issues with the new architecture (see the packages/expo-app which is now configured for the new arch):
> CommandError: Malformed xcodebuild results: "UNLOCALIZED_RESOURCES_FOLDER_PATH" variable was not generated in build output. Please report this issue and run your project with Xcode instead.

If you run into this, clearing everything, DerivedData, .expo, ios/build, android/build, etc, seems to resolve the issue. We are currently attempting to debug this build issue, but wanted to get something out to test and discuss issues that may come up with the new architecture.  Feel free to comment or here or open a new ticket. Thanks!

**Comment by birkskyum** [comment]
> @birkskyum do you think switching to the new arch could be supported by a bounty? 

@KiwiKilian for that, the maplibre-react-native repo has to be elevated first to a [Core tier](https://github.com/maplibre/maplibre/blob/main/PROJECT_TIERS.md#core).

**Comment by KiwiKilian** [comment]
@birkskyum thanks for the feedback. I guess that would need a bigger user base?

**Comment by birkskyum** [comment]
In general yes, if it appear to bring a lot of value to the community, it will be prioritized - OR if we see a need from one or more [sponsor](https://maplibre.org/sponsors/) of MapLibre, that's being paid attention to as well.

**Comment by brentforder** [comment]
> In general yes, if it appear to bring a lot of value to the community, it will be prioritized - OR if we see a need from one or more [sponsor](https://maplibre.org/sponsors/) of MapLibre, that's being paid attention to as well.
> 
> We walk a fine line in balancing our spendings, because the organization only can prioritize investments in community needs, to the extend that we _**at the same time**_ can provide more value to our sponsors, per dollar donated, compared to their alternatives, which historically often have taken the shape of an in-house fork.

I'm honestly surprised that maplibre-react-native isn't more popular, or a core tier. When I saw this option for a multi-platform, open-source geographic map library that works with Expo, I rejoiced that I wouldn't have to mess with native code, native build configurations or inhuman programming languages like Swift. Expo even supports OpenGL ES... And for the cases where you need platform-specific functionality not covered well by Expo (i.e. content sharing between apps), you can generally bootstrap in custom Expo config plugins to maintain that native functionality without too much pain.

**Comment by KiwiKilian** [comment]
I've started the migration process with #828.

# Going all-in on the new Architecture
[RN](https://github.com/reactwg/react-native-new-architecture/discussions/290) and [Expo](https://expo.dev/blog/out-with-the-old-in-with-the-new-architecture#why-migrate) are pushing towards new architecture. So we will do too, from v11 onwards this library will be new architecture only. Those of you who can't upgrade to the new architecture should stick with v10. We have started releasing v11 alpha versions **today**.

## Shameless self-plug
Right now I'm mainly doing the migration in my spare time. If you like what I'm doing and will profit from this migration, consider [sponsoring me on GitHub](https://github.com/sponsors/KiwiKilian) or reaching out to [MapLibre](https://maplibre.org/sponsors/) and **request to funnel the money explicitly to this project**. If you are here for company work, talk to your management 😉.

**Comment by jaltin** [comment]
**Great news, @KiwiKilian!**

I’m in the middle of upgrading our app to the latest React Native, and New Architecture support in maplibre-react-native will be critical for us.

My company (@lineinspector) already sponsors the MapLibre organization. Is there a straightforward way to have part of that contribution earmarked for React Native work?

I’ve also just added a one-time sponsorship for you — thanks so much for driving this effort.

**To everyone else depending on this library:** now is the perfect moment to chip in and help keep the momentum going!

One last question: do you have a rough (even very rough!) timeline for v11? A ballpark would help us plan our own rollout. 😊

Thanks again for all your work!

**Comment by caspg** [comment]
@KiwiKilian That's awesome news! Thanks a lot.

What do you think about creating a sponsorship goal/target? I think I saw something like this in other open source project. Something like I need XXX to fully focus on finishing this migration and so far I have YYY. Also, I read that crowdfundings with clear goals are more successful.

**Comment by KiwiKilian** [comment]
Thanks for those of you who already came forward, highly appreciated!
@jaltin

Are you already using MLRN with old arch or are you migrating from another map framework?

> My company (https://github.com/lineinspector) already sponsors the MapLibre organization. Is there a straightforward way to have part of that contribution earmarked for React Native work?

It's more an idea that initially came by @tyrauber. @birkskyum what could be an approach regarding this topic, while MLRN is not a core project? Nevertheless, this project only exist, because the other core projects exist. Therefore any kind of split should be considered carefully.

> One last question: do you have a rough (even very rough!) timeline for v11? A ballpark would help us plan our own rollout. 😊

I would really hope to have the biggest [issues](https://github.com/maplibre/maplibre-react-native/issues?q=is%3Aissue%20state%3Aopen%20label%3A%22Architecture%3A%20New%22%20type%3ABug) fixed within two to three months. Hopefully v11 released no later then the end of this year, more wished around autumn. TBH this is really wild guess, as the first two month (of spare time, not developer hours) went into the setup, which is sometimes a bit trial and error. So actually code migration can of course go faster or slower, as I've only touched the first module in #828.

@caspg

> What do you think about creating a sponsorship goal/target?

As I'm not yet quite sure the exact hours I will be spending, I find it hard to set a goal on this topic. I also can't/want to pressure myself by setting a goal. I really want to do this migration from an intrinsic motivation. But I figured if people benefit from this work financially, they could channel there benefits here. Which of course also helps the momentum of my work while not entering a contract. But I will see how this develops along the way.

Furthermore, if I get the feeling everything went much faster, there will still be enough to do in [overhauling this library in other ways](https://github.com/maplibre/maplibre-react-native/issues/583#issuecomment-2565298487).

**Comment by jaltin** [comment]
> Are you already using MLRN with old arch or are you migrating from another map framework?

Yes we are using MLRN 9.1.0 with RN 0.73.11. We tried updating to RN 0.76 and MLRN 10 a while back but rolled back as there were too many blockers, and the bugs in new arch  for MLRN 10 was one of the main reasons.

So the faster this can happen, the better it will be for us.

**Comment by KiwiKilian** [comment]
As an intermediate step I can only recommend to upgrade to v10 and sticking with the old arch. For us v10 with old arch is really stable, as we migrated from RNMapbox to this library. It also eases parts of the tooling, like bumping the MapLibre Native version independently of MLRN.

But for sure the migration to new arch is in focus now.

**Comment by jaltin** [comment]
>  For us v10 with old arch is really stable, as we migrated from RNMapbox to this library. 

Yes we are attempting that path right now, but it gives us headaches with other packages that dont support old arch on latest RN (an example is react-native-mmkv). Hopefully we can find fixes for that though.

Looking forward to hear on the progress. And let us know if we can assist with testing or anything else.

**Comment by ice-cap0** [comment]
This will be absolute fire. Can't wait :)

**Comment by jaltin** [comment]
Hi @KiwiKilian, I hope you’re doing well! 👋
Just wanted to check in to see if there’s been any progress behind the scenes on this, or if you have a sense of whether it’s on the horizon or more long-term. Totally understand this is volunteer work — just curious what to expect.

Thanks again for all your efforts on the project! 🙏

**Comment by KiwiKilian** [comment]
I'm making progress on migrating the [`MapView` and `Camera`](https://github.com/maplibre/maplibre-react-native/pull/861). I think their migrations are close to beeing feature complete on both platforms. Next up is testing and fixing bugs (for example normal child `View`s aren't working right now). These two components are the most complex and had quite a few bad API choices. These also contradicted some of the patterns for React Native new Arch, for example it used a single event type to send different kind of events on the `MapView`. For the long term maintenance of this library I choose to cleanup these messy parts and improve those APIs, which will also give much cleaner JS/React API. So I would say finishing these two components is on the horizon, hopefully no later then end of November.

Then I'll have to figure the biggest blockers on other components, as most reported bugs seemed to be on those two components. I'm also trying to get allocated time through my employer, but that's a lengthy process as some customer has to fund it.

I must also note, that this library is rather complex for RN and is hitting some boundaries on React Native new architecture, see https://github.com/facebook/react-native/issues/49920 for example. Resources and documentation on these usecases are really dense and RNMapbox only did the bare minimum to get everything working on the new architecture. So a lot of research, digging in other RN libraries and trial and error was necessary. And I'm also no native dev... I hope this helps to understand, why the migration is a rather lengthy topic, especially because I'm not able to work full time on this.

**Comment by KiwiKilian** [comment]
Quick update to this issue, the migration of `MapView` and `Camera` has been released as `11.0.0-alpha.9`. I've created further steps as sub-issues of this one.

**Comment by KiwiKilian** [comment]
Sorry, didn't want to close the parent task. If anyone is interested to take one of the sub issues you can comment there.

**Comment by RobertSasak** [comment]
Great work! 

If we do not use those 5 missing components(Snapshot, Images, Light...) that are not yet migrated to new architecture, how safe it is to upgrade to alpha version?

**Comment by KiwiKilian** [comment]
There will be a few more breaking changes on those not yet fully migrated components, mainly `Layer`s, `Images` and `SnapshotManager`. Currently still buggy are Markers and Annotations, that should be solved through #1079/#1234 soonish (next week maybe?). If you don't use Markers/Annotations (or using new Arch with v10) you can definitely start trying around the latest alpha.

**Comment by KiwiKilian** [comment]
All components have been migrated to new architecture in the latest alpha! 🚀
